### PR TITLE
Lighthouse Parser Code integrated to work with Hermes Profile

### PR DESCRIPTION
**Pull request by saphal1998**
## Summary

This PR is to convert the Hermes Profile into a format supported by Chrome Dev Tools.

Solves #21

High level breakdown:

* The Hermes Profile generates profiles with the following properties:
1. Trace events - These events are no good, because while they are Chrome Supported, they contain no information about any of the functions run.
2. Samples - There are samples capturing stack frame data at specific instants of time. The app is [sampled](https://www.jetbrains.com/help/profiler/Profiling_Guidelines__Choosing_the_Right_Profiling_Mode.html) and hence we can derive and function call stack information using these `samples`. 
3. Stack Frames - Contains information of individual function calls. These contain function level information of individual events. I believe this [test](https://github.com/GoogleChrome/lighthouse/blob/0422daa9b1b8528dd8436860b153134bd0f959f1/lighthouse-core/test/lib/tracehouse/cpu-profile-model-test.js#L21-L45) explains this best.
* The Hermes Profile needs to be reduced to a `ProfileChunk` so that it can be interpreted by the Lighthouse Parser. This can done by assigning the following to the skeleton of a CPU Profile Chunk
1. pid: All samples contain the same process ID, so can derive this from any of the samples, in this case we use the pid of the first element in the samples array.
2. tid: Javascript being single threaded has a single tid throughout. Hence, this can also be obtained from any of the samples. In this case, we get it from the first sample in the samples array.
3. id: Can be any random string (Not sure about the relevance of ID in our use case.)
4. start time: The start time is the timestamp of the first sample in our array
5. nodes: This can have a 1-to-1 relationship with stackFrames in the Hermes Profile (can be understood through this [test](https://github.com/GoogleChrome/lighthouse/blob/0422daa9b1b8528dd8436860b153134bd0f959f1/lighthouse-core/test/lib/tracehouse/cpu-profile-model-test.js#L21-L45)), wherein the only properties that need proper definition in each individual node are:
    - id: stackFrameId
    - callFrame: contain all properties in the stackFrame
    - parent: parent of a particular stack frame id

6. samples: Can be an array of`sf` numbers from the array of samples.
7. timestamps: The first element is the absolute time stamp of the first sample, while the subsequent elements are time differences between the timestamps of the previous and current samples.

* After having obtained the `Profile Chunk`, this can be parsed by the [Lighthouse Event Parser](https://github.com/GoogleChrome/lighthouse/blob/0422daa9b1b8528dd8436860b153134bd0f959f1/lighthouse-core/lib/tracehouse/cpu-profile-model.js) which converts these Profile Chunks to B/E Duration Events.

## Test Plan

* The profile generated for our sample profiles match the internal tools, however currently we don't have much else to go on.
*  To verify our outputs to some extent, we profile the following code using the hermes sample profiler,
```
const f1 = (n) => {
  for (let i = 0; i < n; i++) {
    Math.random();
  }
};

const fib = (n) => {
  if (n == 0 || n == 1) return n;
  else {
    f1(n)
    return fib(n - 1) + fib(n - 2);
  }
};

const main = (n) => {
  fib(n);
};

main(30);
```
Our transformer yielded these results: 
Screenshot 2020-07-22 at 9 12 55 PM

And for 
```
const f1 = (n) => {
  for (let i = 0; i < n; i++) {
    Math.random();
  }
};

const main = (n) => {
  f1(n);
};

main(30000000);
```
We get, 
Screenshot 2020-07-22 at 9 14 44 PM

And finally for just 
```
const fib = (n) => {
  if (n == 0 || n == 1) return n;
  else {
    return fib(n - 1) + fib(n - 2);
  }
};

const main = (n) => {
  fib(n);
};

main(25);
```

We get,
Screenshot 2020-07-22 at 9 15 39 PM

The reason I think our transformer is getting the order right is because the fibonacci code seems to break the rows into 2 halves at each level which is correct, as we do make a recursive call to fib(n-1) + fib(n-2).  

The timing information is something which is yet to be fixed.

* TODO: Tests need to be written for the transformer.

### CODE REVIEW

**Comment by saphal1998** [comment]
@jessieAnhNguyen I kept those just so we have the same files to test on, will remove them before merge 👍

File: src/index.ts [hunk 1]
```diff
@@ -1,168 +1,43 @@
...
+  );
+}
+
+/* Read Hermes Profile
+ **Replace with CLI transform function as it will just give the path of the CPU profile,
+ **adopting same model here
+ */
+
+const PROFILE_PATH = 'nestedFuncProfile.cpuprofile';
+const DEBUG_SERVER_PORT: string =
```
Comment (jevakallio) [inline src/index.ts:23]: The responsibility of the `hermes-profile-transformer` package should be to export a single function that's able to take in a profile file path and optionally a source map file path, run the transformation, and provide an output as a string. It should not be aware of React Native.

The responsibility of the react-native cli is to provide the inputs.

It's ok for this package to also provide a minimal standalone CLI to run the transformation (when not being used from react-native cli, but that should be in a separate file and provided in the [npm `bin` field](https://docs.npmjs.com/files/package.json#bin). The index file should be just the exported code.

File: src/index.ts [hunk 2]
```diff
@@ -1,168 +1,43 @@
...
+const chromeEvents = profiler.createStartEndEvents();
+
+/**
+ * This function currently takes the input as the RN_SOURCE_MAP URL, however we need to convert this to
+ * take in sourceMap
+ * @see {CpuProfilerModel.changeNamesToSourceMaps}
+ */
+CpuProfilerModel.changeNamesToSourceMaps(RN_SOURCE_MAP, chromeEvents).then(
+  (events: DurationEvent[]) => {
+    writeFileSync(`${PROFILE_PATH.split('.')[0]}.json`, JSON.stringify(events));
```
Comment (jevakallio) [inline src/index.ts:41]: Use `path.basename` instead

File: src/profiler/cpuProfilerModel.ts [hunk 3]
```diff
@@ -0,0 +1,374 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
```
Comment (jevakallio) [inline src/profiler/cpuProfilerModel.ts:2]: Apache 2.0 recommends keeping a notice of changes for original code. If I'm reading this correctly, this file is a mix of Lighthouse code mixed with ours? If that's the case, let's discuss what's the best way to do put that in licensing terms.

Comment (saphal1998) [reply src/profiler/cpuProfilerModel.ts:2]: Yes, essentially I just converted all the code written there into TS and added a couple of functions of my own. Let's continue this on Discord

Comment (jevakallio) [reply src/profiler/cpuProfilerModel.ts:2]: @saphal1998 cool! I think that's fine. In the future, when you source code from elsewhere, it's usually good to layer the commits so that they're easier to review:

- 1/n commit: Just the original file
- 2/n commit: Convert to TypeScript
- 3-n/n commits: Add functionality

That way the code is easier to review (right now I have no idea who's written what), and those commits can be used in the license header to clearly show what changes have been made.

No need to do this retroactively, I'll review 👌

File: src/profiler/cpuProfilerModel.ts [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+    });
+    const stackFrameIds: string[] = Object.keys(stackFrames);
+    // TODO: What should URLs be? Change name to get things from SourceMaps
+
+    const profileNodes: CPUProfileChunkNode[] = stackFrameIds.map(
+      (stackFrameId: string) => {
+        return {
+          id: Number(stackFrameId),
+          callFrame: {
+            line: stackFrames[stackFrameId].line,
```
Comment (jevakallio) [inline src/profiler/cpuProfilerModel.ts:279]: Extract `stackFrames[stackFrameId]` to a variable

File: src/profiler/cpuProfilerModel.ts [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+   * @param {DurationEvent[]} chromeEvents
+   * @throws {Debug Server not running}
+   * @returns {DurationEvent[]}
+   */
+  static async changeNamesToSourceMaps(
+    metroServerURL: string,
+    chromeEvents: DurationEvent[]
+  ): Promise<DurationEvent[]> {
+    /**
+     * To eliminate Metro Server depedency, sourceMap needs to be the input of this function NOT
```
Comment (jevakallio) [inline src/profiler/cpuProfilerModel.ts:324]: 👍

File: src/profiler/cpuProfilerModel.ts [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+  }
+
+  /**
+   * Refer to the source maps for the `index.bundle` file. Throws error if debug server not running
+   * @param {string} metroServerURL
+   * @param {DurationEvent[]} chromeEvents
+   * @throws {Debug Server not running}
+   * @returns {DurationEvent[]}
+   */
+  static async changeNamesToSourceMaps(
```
Comment (jevakallio) [inline src/profiler/cpuProfilerModel.ts:319]: Does this function need to live on this class at all? Its seems unrelated to the main purpose of this class, and doesn't appear to use any internal state?

I'd move it into its own function that gets called after the main profile transformation happens.
